### PR TITLE
Generate ".d.ts" definition files for Stylable stylesheets

### LINKED ISSUE #351: Feature proposal: Live DTS generation
> Note: this proposal used to exist as a PR that is now closed, re-opening as issue.

# Live DTS generation

We love TypeScript and we use it extensively for its Strong-Typed dev experience.

In order to get better completions and validations for our TypeScript components we want to add a DTS generator to Stylable. If you elect not to use it all your styleSheets will default to a less strong-typed interface.

By using the DTS-generator, a `*.st.css.d.ts` file will be generated for each `*.st.css` file in your project in its directory.

This allows for exact completions on the `classes` and `states` applied from the stylesheet.

# Example
CSS:
```css
/* my-comp.st.css */
.myInnerPart{
    -st-states:index(number);
    color:red;
}
```

Stylable runtime DTS (imported in the resulting file):
```ts
    type StateValue = boolean | number | string;

    interface StateMap extends NativeStateMap { // extending native css states, e.g. :hover
        [stateName: string]: StateValue;
    }

    interface StyleSheet{
        <ClassName extends ClassNames>( 
            className: ClassName | ClassName[], 
            states?: StateMap[ClassName] ): string;
    }
    declare const runtime: StyleSheet
    export default runtime
```

generated DTS:
```ts
type ClassNames =  "myInnerPart" | "root"
type StateMap = {
    myInnerPart: {
        selected: number
    }
}

const stylesheet: import('stylable-runtime')
    .IRuntimeStylesheet<ClassNames, StateMap>;

export default stylesheet;
var s = {} as StyleSheet;
```

> Note by @barak007: this example does not take `props` into account, ignoring `classNames` and `states` they may introduce to the stylesheet.

### PR DESCRIPTION
**Pull request by tomrav**
By generating a `.d.ts` file for every component we aim to improve the type checking of components and how they use the inner parts of stylesheets.

These `.d.ts` files can validate the following:
- `classes` - check that used classes exist on the imported stylesheet
- `states` - and their values are checked against the class that defined them
- `variable/keyframes` - are checked to exist in the stylesheet
- `stVars` - checked that they exist and display value generated in build-time

Execution flow will be using CLI in watch mode, and adding the generated file to the project `.gitignore`. These should be published alongside any `.st.css` file.

Tasks:
- [x] Generate `.d.ts` for better validation and completion when consuming `.st.css` stylesheets in TypeScript
- [x] Implement source-maps for the `.d.ts` files to allow ctrl+click to source (`.st.css`)
- [x] Expose functionality through the CLI
- [x] Narrow the `StylableDOMUtil` to not require the `cssStates` function, but to instead use stylesheet namespace and generate states itself
- [x] Add docs - wixplosives/stylable.io#22

~~Blocked by #1743~~
Resolves #351

### CODE REVIEW

**Comment by barak007** [comment]
TODO: 

- [ ] Move generate dts to module-utils
- [ ] Add Typescript e2e validation tests
- [ ] Research source map support to allow Ctrl+click to source
- [ ] Maybe change the StylableDOMUtil to not require cssStates but to use the namespace and generate them itself
- [ ] Add Docs

**Comment by idoros** [comment]
> I didn't saw any behavior change with or without it. maybe for correctness

without is you can pass anything as the first argument 🤨

**Comment by idoros** [comment]
closes: #351

File: packages/module-utils/src/generate-dts-sourcemaps.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+export function generateDTSSourceMap(dtsContent: string, meta: StylableMeta) {
+    const tokens = tokenizeDTS(dtsContent);
+    const mapping: Record<number, LineMapping> = {};
+    const lines = dtsContent.split('\n');
+    let lastSrcLocation: Position = { line: 0, column: 0 };
+
+    for (const dtsLine of lines.keys()) {
+        const resToken = findTokenForLine(dtsLine, tokens);
```
Comment (idoros) [inline packages/module-utils/src/generate-dts-sourcemaps.ts:191]: this makes an assumption of the way the `d.ts` is structured (1 symbol at a line) - maybe add a comment explaining how this function/util works

File: packages/module-utils/src/generate-dts-sourcemaps.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+                    );
+                    lastSrcLocation = { ...currentSrcPosition };
+                }
+            } else if (resToken.type === 'states') {
+                const classSourceName = getClassSourceName(
+                    resToken.token.className.value,
+                    tokens[1] as ClassesToken
+                );
+
+                if (classSourceName) {
```
Comment (idoros) [inline packages/module-utils/src/generate-dts-sourcemaps.ts:226]: is there a valid source that produce state in a class that cannot be found? maybe raise a warning?

- same question for other symbols with source position that cannot be located

File: packages/cli/src/build.ts [hunk 3]
```diff
@@ -198,7 +206,28 @@ function buildSingleFile(
...
+
+        log('[Build]', 'output .d.ts');
+        tryRun(
+            () => fs.writeFileSync(outSrcPath + '.d.ts', dtsContent),
+            `Write File Error: ${outPath}`
+        );
+
+        // .d.ts.map
+        // if not explicitly defined, assumed true with "--dts" parent scope
+        if (dtsSourceMap === undefined || dtsSourceMap === true) {
```
Comment (idoros) [inline packages/cli/src/build.ts:221]:
```suggestion
        if (dtsSourceMap !== false) {
```

File: packages/cli/src/build.ts [hunk 4]
```diff
@@ -198,7 +206,28 @@ function buildSingleFile(
...
+            const dtsMappingContent = createModuleSource(res, 'dts.map', false);
```
Comment (idoros) [inline packages/cli/src/build.ts:222]: if `createModuleSource` would accept options or would provide separate functions for each module type, then I would prefer passing the created `d.ts` rather then recreating it internally

File: packages/module-utils/test/generate-dts.spec.ts [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+            'test.st.css': '.root{}',
+            'test.ts': `
+                import { eq } from "./test-kit";
+                import { classes } from "./test.st.css";
+                
+                eq<"test__root">(classes.root);
+            `,
+        });
+
+        expect(tk.typecheck('test.ts')).to.equal('');
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/generate-dts.spec.ts:28]: consider also adding syntax error checking

File: packages/module-utils/test/generate-dts.spec.ts [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    });
+
+    it('should generate Stylable var .d.ts', () => {
+        tk.populate({
+            'test.st.css': ':vars {c1: green;}',
+            'test.ts': `
+                import { eq } from "./test-kit";
+                import { stVars } from "./test.st.css";
+                
+                eq<string>(stVars.unknown);
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/generate-dts.spec.ts:80]: should be c1

File: packages/module-utils/test/generate-dts.spec.ts [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+                'test.st.css': '.root {}',
+                'test.ts': `
+                    import { st, classes } from "./test.st.css";
+                    
+                    st(classes.root, { state1: true });
+                `,
+            });
+
+            expect(tk.typecheck('test.ts')).to.include(
+                "Argument of type '{ state1: boolean; }' is not assignable"
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/generate-dts.spec.ts:210]: this may be a bit fragile, let's fix it up when it breaks for the first time

File: packages/module-utils/test/generate-dts.spec.ts [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+            );
+        });
+
+        it('should warn on type mis-match in state', () => {
+            tk.populate({
+                'test.st.css': '.root {}',
+                'test.ts': `
+                    import { st, classes } from "./test.st.css";
+                    
+                    st(classes.root, { state1: 'str' });
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/generate-dts.spec.ts:220]: source is wrong here, should define state with different type

File: packages/module-utils/src/generate-dts.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+ * support custom types?
+ */
+function getStateTSType(stateDef: StateParsedValue | null) {
+    return stateDef === null
+        ? 'boolean'
+        : stateDef.type === 'enum'
+        ? stateDef.arguments
+              .map((v) => (typeof v === 'string' ? asString(v) : 'unknown'))
+              .join(' | ')
+        : stateDef.type /* what are the rest of the states */;
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/src/generate-dts.ts:49]: need to remove comment

File: packages/cli/src/build.ts (hunk 4 above)
Comment (tomrav) [reply packages/cli/src/build.ts:222]: will open issue

File: packages/module-utils/src/index.ts [hunk 10]
```diff
@@ -1,2 +1,5 @@
 export * from './module-source';
 export * from './module-factory';
+export * from './generate-dts';
+export * from './dts-rough-tokenizer';
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/src/index.ts:4]: agreed

File: packages/cli/src/build.ts (hunk 3 above)
Comment (tomrav) [reply packages/cli/src/build.ts:221]: fixing

File: packages/module-utils/src/generate-dts-sourcemaps.ts (hunk 2 above)
Comment (tomrav) [reply packages/module-utils/src/generate-dts-sourcemaps.ts:226]: this is mostly due to typing concerns, can't really think of a legit case

File: packages/module-utils/src/generate-dts.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+function stringifyStringRecord(record: Record<string, string>, indent = SPACING) {
+    return Object.keys(record)
+        .map((k) => `${indent}${asString(k)}: string;`)
+        .join('\n');
+}
+
+function stringifyClasses(classes: Record<string, string>, namespace: string, indent = SPACING) {
+    return Object.keys(classes)
+        .map((name) => `${indent}${asString(name)}: ${asString(scope(name, namespace))};`)
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/src/generate-dts.ts:14]: If you're building your stylesheets with a custom namespace resolver, I think you can reach the suggested setup.
A future build may change the already resolved scoped classes (when building from the .st.css).

I added a comment.

File: packages/module-utils/test/dts-rough-tokenizer.spec.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+    beforeEach(async () => {
+        tk = await new DTSKit().init();
+    });
+
+    afterEach(async () => {
+        await tk.dispose();
+    });
+
+    it('should tokenize simple states', () => {
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/dts-rough-tokenizer.spec.ts:16]: I think it's a decent start considering it's covered by more e2e sort of tests, we can add more if we come across issues.

File: packages/module-utils/test/dts-rough-tokenizer.spec.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        const states = getLocalClassStates('root', out);
+
+        expect(states[0].stateName.value, 'stateName').to.equal('"myState"');
+        expect(states[0].type[0].value, 'stateValue Type').to.equal('boolean');
+
+        expect(states[1].stateName.value, 'stateName2').to.equal('"myState2"');
+        expect(states[1].type[0].value, 'stateValue Type string').to.equal('string');
+
+        expect(states[2].stateName.value, 'stateName3').to.equal('"myState3"');
+        expect(states[2].type[0].value, 'stateValue Type enum a').to.equal('"a"');
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/dts-rough-tokenizer.spec.ts:33]: shouldn't really matter in our very specific case.

File: packages/module-utils/test/generate-dts.spec.ts [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+                `,
+            });
+
+            expect(tk.typecheck('test.ts')).to.include(
+                "Argument of type '{ state1: string; }' is not assignable to parameter of type"
+            );
+        });
+    });
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/generate-dts.spec.ts:228]: added.

File: packages/module-utils/src/index.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (idoros) [reply packages/module-utils/src/index.ts:4]: still exported

File: packages/module-utils/test/generate-dts.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/generate-dts.spec.ts:28]: done both

File: packages/module-utils/test/test-kit/dts-kit.ts [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+export class DTSKit {
+    tmp!: ITempDirectory;
+    populate!: (files: Record<string, string>) => void;
+    stylable!: Stylable;
+    sourcePath!: (filePath: string) => string;
+    write!: (internalPath: string, content: string) => void;
+    read!: (internalPath: string) => string;
+    genDTS!: (internalPath: string) => void;
+    typecheck!: (internalPath: string) => string;
+    async init() {
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/test/test-kit/dts-kit.ts:23]: cleaned this up

File: packages/cli/src/build.ts (hunk 4 above)
Comment (tomrav) [reply packages/cli/src/build.ts]: opened #1843

File: packages/cli/README.md [hunk 16]
```diff
@@ -15,31 +15,33 @@ This package is currently a work-in-progress and is set to `private`. Once it ma
...
+| `--indexFile`             |        | filename of the generated index                                                        | `false`                                      |
+| `--cjs`                   |        | output commonjs modules (`.js`)                                                        | `true`                                       |
+| `--esm`                   |        | output esm modules (`.mjs`)                                                            | `false`                                      |
+| `--css`                   |        | output transpiled css files (`.css`)                                                   | `false`                                      |
+| `--stcss`                 |        | output stylable source files (`.st.css`)                                               | `false`                                      |
+| `--dts`                   |        | output definition files for the stylable source files (`.st.css.d.ts`)                 | `false`                                      |
```
Comment (barak007) [inline packages/cli/README.md:29]: I think we need dts location option that is not related to the outDir

File: packages/cli/src/build.ts [hunk 17]
```diff
@@ -198,7 +206,39 @@ function buildSingleFile(
...
+        log('[Build]', 'output .d.ts');
+        tryRun(
+            () => fs.writeFileSync(outSrcPath + '.d.ts', dtsContent),
+            `Write File Error: ${outPath}`
+        );
+
+        // .d.ts.map
+        // if not explicitly defined, assumed true with "--dts" parent scope
+        if (dtsSourceMap !== false) {
+            const dtsMappingContent = createModuleSource(
```
Comment (barak007) [inline packages/cli/src/build.ts:222]: please don't use `createModuleSource` it wasn't a good stab for generating dts.map. I don't like the undefined

File: packages/cli/test/cli.spec.ts [hunk 18]
```diff
@@ -184,6 +184,72 @@ describe('Stylable Cli', () => {
...
+        const stylesheetContent = dirContent[join('dist', 'style.st.css')];
+        const dtsContent = dirContent[join('dist', 'style.st.css.d.ts')];
+        const dtsSourceMapContent = dirContent[join('dist', 'style.st.css.d.ts.map')];
+
+        expect(stylesheetContent).to.equal(srcContent);
+        expect(dtsContent.startsWith('/* THIS FILE IS AUTO GENERATED DO NOT MODIFY */')).to.equal(
+            true
+        );
+        expect(
+            dtsSourceMapContent.startsWith('{\n    "version": 3,\n    "file": "style.st.css.d.ts"')
```
Comment (barak007) [inline packages/cli/test/cli.spec.ts:216]: not sure that the version in important to check here.

File: packages/dom-test-kit/src/stylable-dom-util.ts [hunk 19]
```diff
@@ -14,8 +14,16 @@ export interface PartialElement {
     classList: Element['classList'];
 }
 
+export type StylesheetHost = {
```
Comment (barak007) [inline packages/dom-test-kit/src/stylable-dom-util.ts:17]: is this used any where now?

File: packages/module-utils/src/generate-dts.ts [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+
+declare const classes: {${classes}};
+
+declare const vars: {${vars}};
+
+declare const stVars: {${stVars}};
+
+declare const keyframes: {${keyframes}};
+
+declare function st<T extends string = keyof states>(
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/module-utils/src/generate-dts.ts:83]: it's true but it will make the dts less standalone and will require others to install `@stylable/runtime` if published.
I'm not sure what is best here.

File: packages/cli/README.md (hunk 16 above)
Comment (barak007) [reply packages/cli/README.md:29]: I mean that we need to provide a way if someone wants to build the d.ts into dist vs src

File: packages/module-utils/src/generate-dts.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/module-utils/src/generate-dts.ts:83]: That's what's been holding me back on turning it into an import. I think for now we'll leave it as is.

File: packages/cli/README.md (hunk 16 above)
Comment (tomrav) [reply packages/cli/README.md:29]: if we add the option, will it generate them only there and not in the output dir? we may want it to output to both source and target so that they are compatible in their content

File: packages/cli/test/cli.spec.ts (hunk 18 above)
Comment (tomrav) [reply packages/cli/test/cli.spec.ts:216]: Just checking the start of the sourcemap, the version here is for the sourcemap syntax.

File: packages/dom-test-kit/src/stylable-dom-util.ts (hunk 19 above)
Comment (tomrav) [reply packages/dom-test-kit/src/stylable-dom-util.ts:17]: I think not, should I fix (and use) or remove this?

File: packages/cli/README.md (hunk 16 above)
Comment (tomrav) [reply packages/cli/README.md:29]: perhaps an option like, `srcDir | outDir | both` would work. I think we'll start without this, and see what real-world cases behave like.

File: packages/cli/src/build.ts (hunk 17 above)
Comment (tomrav) [reply packages/cli/src/build.ts:222]: Will extract.

File: packages/dom-test-kit/src/stylable-dom-util.ts (hunk 19 above)
Comment (tomrav) [reply packages/dom-test-kit/src/stylable-dom-util.ts:17]: Choose to fix.

File: packages/eslint-plugin-stylable/README.md [hunk 21]
```diff
@@ -40,3 +40,9 @@ or
 |-----------|:-----:|:-------:|-------------|
 | `resolveOptions` | `object` | `{}` | Stylable resolver options |
 | `exposeDiagnosticsReports` | `boolean` | `false` | expose Stylable transpilation diagnostics on the import statement |
+
+## Known issue
+
+ESLint is not intended for multi-file operations, such as validating your TS when an `.st.css` files changes. This means that the plugin can slip out of sync and show stale data.
```
Comment (idoros) [inline packages/eslint-plugin-stylable/README.md:46]:
```suggestion
ESLint is not intended for multi-file operations, such as validating your TS when an `.st.css` files changes. This means that the plugin might lose track of stylesheet changes and work with stale data.
```

File: packages/eslint-plugin-stylable/README.md [hunk 22]
```diff
@@ -40,3 +40,9 @@ or
...
+
+This behavior will show out-of-date errors in your TS files, and only manually triggering a change in that file will remove the error.
```
Comment (idoros) [inline packages/eslint-plugin-stylable/README.md:48]:
```suggestion
This behavior will show out-of-date errors or will miss new errors in your TS files, and only manually triggering a change in the stylesheet will sync ESlint and update the diagnostics.
```
